ethcore/src/engines/clique/block_state.rs

```diff
@@ -0,0 +1,207 @@
+
+		if (inturn && *header.difficulty() != DIFF_INTURN) ||
+			(!inturn && *header.difficulty() != DIFF_NOTURN) {
+			return Err(From::from(format!("Error applying #{}({}): wrong difficulty!",
+											header.number(),
+											header.hash())));
+		}
+
+		Ok(creator)
+	}
+
+	/// Verify and apply an new header to current state, might fail with error.
+	pub fn apply(&mut self, header: &Header, is_checkpoint: bool) -> Result<Address, Error> {
+		let creator = self.verify(header)?;
+
+		// rotate recent signers.
+		self.recent_signers.push_front(creator.clone());
+		if self.recent_signers.len() >= ( self.signers.len() / 2 ) + 1 {
+			self.recent_signers.pop_back();
+		}
+
+		if is_checkpoint {
+			// checkpoint block should not affect previous tallying, so we check that.
+			let signers = extract_signers(header)?;
+			if self.signers != signers {
+				return Err(From::from("checkpoint block signers is different than expected"));
+			};
+
+			self.votes.clear();
+			self.votes_history.clear();
+
+			// maybe release some memory.
+			self.votes.shrink_to_fit();
+			self.votes_history.shrink_to_fit();
+
+			return Ok(creator);
+		}
+
+		let beneficiary = *header.author();
+
+		// No vote, ignore.
+		if beneficiary == NULL_AUTHOR {
+			return Ok(creator);
+		}
+
+		let nonce = *header.decode_seal::<Vec<&[u8]>>()?.get(1).ok_or(
+			"Error decoding seal"
+		)?;
+
+		let vote_type: VoteType;
+
+		if NONCE_AUTH_VOTE == nonce {
+			vote_type = VoteType::Add;
+		} else if NONCE_DROP_VOTE == nonce {
+			vote_type = VoteType::Remove;
+		} else {
+			return Err(From::from("beneficiary specified but nonce was not AUTH or DROP"));
+		};
+
+		// Record this vote, also since we are using an hashmap, it will override previous vote.
+		self.votes.insert((creator.clone(), beneficiary), vote_type);
+		self.votes_history.push((header.number(), creator.clone(), vote_type, beneficiary));
+
+		// Tally up current target votes.
+		let threshold = self.signers.len() / 2;
+		let vote = self.votes.iter().filter(|(key, value)| {
+			(**key).1 == beneficiary && **value == vote_type
+		}).count();
+
+		if vote > threshold {
+			match vote_type {
+				VoteType::Add => {
+					self.signers.push(beneficiary);
+				},
+				VoteType::Remove => {
+					let pos = self.signers.binary_search(&beneficiary);
+					if pos.is_ok() {
+						self.signers.remove(pos.unwrap());
+					}
+				}
+			}
+
+			self.signers.sort();
+
+			// Remove all votes about or made by this beneficiary
+			{
+				let votes_copy = self.votes.clone();
+				let items: Vec<_> = votes_copy.iter().filter(|(key, _value)| {
+					(**key).0 == beneficiary || (**key).1 == beneficiary
+				}).collect();
+
+				for (key, _) in items {
+					self.votes.remove(&key);
+				}
+			}
+		}
+
+		// No cascading votes.
+		Ok(creator)
+	}
+
+	pub fn calc_next_timestamp(&mut self, header: &Header, period: u64) {
+		let base_time = UNIX_EPOCH + Duration::from_secs(header.timestamp());
```

Comment:
This will panic if `Duration` > `libc::time_t::max_value` (it is a signed value usally i32 or i64 AFAIK)

Can you use this the helper function [timestamp_checked_add](https://github.com/paritytech/parity-ethereum/blob/master/ethcore/src/verification/verification.rs#L51) in these cases?

Reply:
Then you should have to change the interface a bit and return a value indicating success or not

Reply:
ref #10383